File: models/naive_bayes.py
import pandas as pd
import numpy as np
from sklearn.naive_bayes import GaussianNB

import os, sys
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname('.'), '..')))
from context import CrossValidation
from ModelFramework.CrossValidation.python.cross_validation import CrossValidation

import importlib
loader = importlib.machinery.SourceFileLoader('report', '/full/path/report/other_py_file.py')
handle = loader.load_module('report')

header = pd.read_csv('data/train.csv', nrows=0)
data = pd.read_csv('data/train_sample_75k.csv', header=None, names=header.columns)

model = GaussianNB()
X = pd.get_dummies(data[["Producto_ID"]], columns=["Producto_ID"], sparse=True)
model.fit(X.ix[:70000, ], np.array(data.ix[:70000, 'Demanda_uni_equil']))
preds = model.predict(X.ix[70000:, ])

File: models/naive_bayes.R
library(data.table)
library(glmnet)
library(biglm)
options(scipen=500)

sapply(dir("../ModelFramework/CrossValidation/crossValidation/R/", full.names = TRUE),
       source)
source("R/error.R")
source("../ModelFramework/Models/naiveBayesRegression.R")

# train = fread("data/train.csv")
# test = fread("data/test.csv")

# train = fread("data/train_subset_of_features.csv")
# droppedCols = c("Town", "State", "has_multigrain", "has_vanilla", "has_choco", "weight_per_piece", "brand", "weight",
#                 "Venta_hoy", "Dev_uni_proxima", "Dev_proxima")
# train[, c(droppedCols) := NULL]
# test = fread("data/test_subset_of_features.csv")
# test[, c(droppedCols) := NULL]
# gc()

train = fread("data/train_complete_45k_features.csv")
test = fread("data/test_1000_features.csv")


modelCols = c("Town", "Canal_ID", "brand", "has_choco", "Cliente_ID", "Producto_ID",
              "Ruta_SAK", "has_vanilla", "has_multigrain")
for(col in modelCols){
    train[, c(col) := as.factor(get(col))]
    test[, c(col) := as.factor(get(col))]
}
predModel = function(model, X){
    out = predictNaiveBayesRegression(model, X)
    out[out < 0] = 0
    return(out)
}
fitModel = function(X, y){
    fitNaiveBayesRegression(X[, modelCols, with=FALSE], y=X$Demanda_uni_equil,
                            log=FALSE)
}
productoAvg = crossValidation(model=list(predict=predModel, fit=fitModel),
                              xTrain = train,
                              yTrain = train[, Demanda_uni_equil],
                              xTest = test,
                              cvIndices = train[, Semana],
                              cvTime=TRUE)
summary(productoAvg)
run(productoAvg, filename="naive_bayes", metric=rmsle, logged=FALSE, plotResults=FALSE)
